subsys/bluetooth/host/adv.c

```diff
@@ -524,27 +524,34 @@ static int hci_set_ad(uint16_t hci_op, const struct bt_ad *ad, size_t ad_len)
 }
 
 static int hci_set_adv_ext_complete(struct bt_le_ext_adv *adv, uint16_t hci_op,
-				    const struct bt_ad *ad, size_t ad_len)
+				    size_t total_len, const struct bt_ad *ad, size_t ad_len)
 {
 	struct bt_hci_cp_le_set_ext_adv_data *set_data;
 	struct net_buf *buf;
 	int err;
 	uint8_t max_data_size;
 
-	buf = bt_hci_cmd_create(hci_op, sizeof(*set_data));
-	if (!buf) {
-		return -ENOBUFS;
-	}
-
-	set_data = net_buf_add(buf, sizeof(*set_data));
-	(void)memset(set_data, 0, sizeof(*set_data));
-
 	if (atomic_test_bit(adv->flags, BT_ADV_EXT_ADV)) {
 		max_data_size = BT_HCI_LE_EXT_ADV_FRAG_MAX_LEN;
 	} else {
 		max_data_size = BT_GAP_ADV_MAX_ADV_DATA_LEN;
 	}
 
+	if (total_len > max_data_size) {
+		return -EINVAL;
+	}
+
+	max_data_size = total_len;
+	total_len += offsetof(struct bt_hci_cp_le_set_ext_adv_data, data);
```

Comment:
I would actually prefer to create a new `const` variable which is something like
```c
const size_t cmd_size = total_len + offsetof(struct bt_hci_cp_le_set_ext_adv_data, data);
```

and use that instead of incrementing the `total_len`. `total_len` should then also become `total_data_len`. 

Alternative instead of using `offsetof` is to modify the struct from 
```c
struct bt_hci_cp_le_set_ext_adv_data {
	uint8_t  handle;
	uint8_t  op;
	uint8_t  frag_pref;
	uint8_t  len;
	uint8_t  data[251];
} __packed;
```
to
```c
struct bt_hci_cp_le_set_ext_adv_data {
	uint8_t  handle;
	uint8_t  op;
	uint8_t  frag_pref;
	uint8_t  len;
	uint8_t  data[0];
} __packed;
```
and then in all use cases when creating the cmd use something like
```c
bt_hci_cmd_create(BT_HCI_OP_LE_SET_EXT_ADV_DATA, sizeof(struct bt_hci_cp_le_set_ext_adv_data) + data_len; 
```

which we do for many other dynamic sized HCI commands.

Reply:
I prefer to use `offsetof`, because the latter needs to be changed at a larger point and has more uncertainty.